src/v/utils/process.cc

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "utils/process.h"
+
+#include "ssx/future-util.h"
+#include "vassert.h"
+#include "vlog.h"
+
+#include <seastar/core/loop.hh>
+#include <seastar/core/when_all.hh>
+#include <seastar/util/log.hh>
+
+#include <sstream>
+#include <string>
+
+static ss::logger proclog{"process"};
+
+namespace ssx {
+
+ss::future<>
+consume_input_stream(ss::input_stream<char> stream, bool is_stdout) {
+    ss::sstring stream_name{is_stdout ? "STDOUT" : "STDERR"};
+
+    while (!stream.eof()) {
+        auto buf = co_await stream.read();
+        if (buf.empty()) {
+            continue;
+        }
+
+        std::string str_buf{buf.begin(), buf.end()}, line;
+        std::stringstream ss(str_buf);
+        while (!ss.eof()) {
+            std::getline(ss, line);
+            vlog(proclog.trace, "{}: {}", stream_name, line);
+        }
+    }
+}
+
+process::~process() {
+    if (is_running()) {
+        vlog(
+          proclog.trace,
+          "Process still running, sending SIGKILL: cmd {}",
+          _cmd_str);
+        try {
+            _process->kill();
+        } catch (const std::system_error&) {
+            // If SIGKILL failed, then the process was already dead and it is
+            // safe to silently ignore.
+        }
+    }
+}
+
+ss::future<std::error_code> process::spawn(
+  const std::filesystem::path& cmd, ss::experimental::spawn_parameters params) {
+    gate_guard g{_gate};
+    if (is_running()) {
+        vlog(proclog.error, "A process is already running: cmd {}", _cmd_str);
+        co_return process_errc::running;
+    }
+
+    // Set command string for logging purposes
+    _cmd_str = cmd.string();
+
+    auto p = co_await ss::experimental::spawn_process(cmd, std::move(params));
+    _process.emplace(std::move(p));
+
+    // Capture output async in the background so the broker is not blocked.
+    ssx::background = ssx::spawn_with_gate_then(_gate, [this] {
+        return consume_input_stream(_process->stdout(), true)
+          .then(
+            [this] { return consume_input_stream(_process->stderr(), false); })
+          .then([this] { return handle_wait(); })
+          .finally([this] { _process.reset(); });
```

Comment:
It it intended that spawn can be used multiple times on a process? If so I fear this could race.

Reply:
Test the concurrency some more please.

Reply:
Two or more process can not run on the same instance of `ssx::process` and the method `ssx::process::spawn` has guards against this as well. I have also extended the sigterm and sigkill unit tests to check `ssx::process::spawn` after terminate.

Finally, I ran the updated sigterm unit test locally for 1 hour (1775 runs) and they all succeeded.